src/Microsoft.AspNetCore.HttpsPolicy/HttpsRedirectionMiddleware.cs

```diff
@@ -0,0 +1,114 @@
+        /// </summary>
+        /// <param name="next"></param>
+        /// <param name="serverAddressesFeature">The</param>
+        /// <param name="options"></param>
+        public HttpsRedirectionMiddleware(RequestDelegate next, IServerAddressesFeature serverAddressesFeature, IOptions<HttpsRedirectionOptions> options)
+        {
+            if (next == null)
+            {
+                throw new ArgumentNullException(nameof(next));
+            }
+
+            if (options == null)
+            {
+                throw new ArgumentNullException(nameof(options));
+            }
+            _next = next;
+
+            var httpsRedirectionOptions = options.Value;
+            _httpsPort = httpsRedirectionOptions.HttpsPort;
+            _statusCode = httpsRedirectionOptions.RedirectStatusCode;
+            _serverAddressesFeature = serverAddressesFeature;
+        }
+
+        /// <summary>
+        /// Invokes the HttpsRedirectionMiddleware
+        /// </summary>
+        /// <param name="context"></param>
+        /// <returns></returns>
+        public Task Invoke(HttpContext context)
+        {
+            if (!context.Request.IsHttps)
+            {
+                if (!_evaluatedServerAddressesFeature && _httpsPort == null)
+                {
+                    CheckAddressesFeatureForHttpsPorts();
+                }
+                _evaluatedServerAddressesFeature = true;
+
+                var host = context.Request.Host;
+                if (_httpsPort.HasValue && _httpsPort.Value > 0)
+                {
+                    host = new HostString(host.Host, _httpsPort.Value);
+                }
+                else
+                {
+                    host = new HostString(host.Host);
+                }
+
+                var request = context.Request;
+                var redirectUrl = UriHelper.BuildAbsolute("https", 
+                                                        host,
+                                                        request.PathBase,
+                                                        request.Path,
+                                                        request.QueryString);
+
+                context.Response.StatusCode = _statusCode;
+                context.Response.Headers[HeaderNames.Location] = redirectUrl;
+                return Task.CompletedTask;
+            }
+
+            return _next(context);
+        }
+
+        private void CheckAddressesFeatureForHttpsPorts()
+        {
+            // The IServerAddressesFeature will not be ready until the middleware is Invoked,
+            int? httpsPort = null;
+            foreach (var address in _serverAddressesFeature.Addresses)
+            {
+                if (Uri.TryCreate(address, UriKind.Absolute, out var uri))
+                {
+                    if (uri.Scheme == "https")
+                    {
+                        // If we find multiple https ports specified, throw
+                        if (httpsPort != null)
```

Comment:
allow multiple if the ports match. E.g. if the ports are all X then it will be OK.